Google.Api.Gax.Grpc/Rest/PartialDecodingStreamReader.cs

```diff
@@ -43,117 +101,147 @@ public PartialDecodingStreamReader(Task<TextReader> textReaderTask, Func<string,
-                    JObject.Parse(currentStr);
+                    _errorMessage = "Response stream completed without a closing array";
                 }
-                catch (Newtonsoft.Json.JsonReaderException)
+                return;
+            }
+
+            // Note: we process the whole of the buffer, even if we see end-of-response or end-of-stream.
+            for (int i = 0; i < charsRead; i++)
+            {
+                char c = _readBuffer[i];
+                var nextAction = _jsonState.Push(c);
+                switch (nextAction)
                 {
-                    // Tried to parse a partial json because the `}` was a part of
-                    // a string or a child inner object.
-                    continue;
+                    case ParseResponse:
+                        _currentResponseBuffer.Append(c);
+                        _queuedResponses.Enqueue(_responseConverter(_currentResponseBuffer.ToString()));
+                        _currentResponseBuffer.Clear();
+                        break;
+                    case SignalEndOfResponses:
+                        // We remember that we've seen the "end of responses" but continue to read
+                        // until the end of the data, to spot badly-behaved responses that include
+                        // data after the closing array. The state tracker will just return
+                        // IgnoreAndContinue or SignalError from here onwards.
+                        _completed = true;
+                        break;
+                    case SignalError:
+                        // Note: we don't have any more information about the way in which the JSON
+                        // was invalid, but we really don't expect to see this anyway.
+                        _errorMessage = "Invalid JSON response.";
+                        // No need to read any more: we're in a failure mode now, other than
+                        // returning any already-parsed responses.
+                        return;
+                    case IgnoreAndContinue:
+                        break;
+                    case BufferAndContinue:
+                        _currentResponseBuffer.Append(c);
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Bug in GAX support library: unhandled state: {nextAction}");
                 }
-
-                TResponse responseElement = _responseConverter(currentStr);
-                _readyResults.Enqueue(responseElement);
-                _currentBuffer.Clear();
             }
         }
-
-        if (_readyResults.Count > 0)
-        {
-            Current = _readyResults.Dequeue();
-            return true;
-        }
-        else
+        // If *anything* fails, this signals an overall failure.
+        // Let the calling code effectively rethrow. (We lose the stack trace, which isn't ideal, admittedly.)
```

Comment:
I think with my comments above, we can just dispose here, and rethrow without loosing the stack trace.
Basically, if we dispose on error, we don't need to save the _errorMessage.

Reply:
We need to save *something*, because if the caller calls MoveNext again, they should get the exception again - potentially *unless* the previous failure was due to cancellation, and they use a different cancellation token the next time. (That's a really tricky case, as we may have lost data. Ouch.)

Reply:
For completeness we discussed this offline and we agreed that it's OK that the enumerator does not recover after a failed MoveNext, even if because cancellation.